src/main/kotlin/org/jitsi/nlj/dtls/DtlsStack.kt

```diff
@@ -199,58 +153,52 @@ class DtlsStack(
-        val packet = packetInfo.packet
-        System.arraycopy(packet.buffer, packet.offset, buf, off, packet.length)
-
-        BufferPool.returnBuffer(packetInfo.packet.buffer)
+    private fun verifyAndValidateRemoteCertificate(remoteCertificate: Certificate?) {
+        remoteCertificate?.let {
+            DtlsUtils.verifyAndValidateCertificate(it, remoteFingerprints)
+            // The above throws an exception if the checks fail.
+            logger.cdebug { "Fingerprints verified." }
+        }
+    }
 
-        return packet.length
+    fun sendApplicationData(data: ByteArray, off: Int, len: Int) {
+        dtlsTransport?.send(data, off, len)
     }
 
     /**
-     * Send an outgoing DTLS packet (already processed by the DTLS stack) out via invoking
-     * a handler.
+     * We get 'pushed' the data from a lower transport layer, but bouncycastle wants to 'pull' the data
+     * itself.  To mimic this, we put the received data into a queue, and then 'pull' it through ourselves by
+     * calling 'receive' on the negotiated [DTLSTransport].
      *
-     * We have to use a synchronous callback approach here, as some packets originate
-     * from within the stack itself (e.g. during connect) and if we put the packets in,
-     * for example, a queue, we'd still have to fire some trigger for something to come
-     * in and read them.
+     * Note: the data we get here may be a DTLS protocol packet and therefore won't generate anything
+     * to be received by [dtlsTransport] or a DTLS app packet (data sent over DTLS after the handshake has
+     * completed) and will result in data being received through [dtlsTransport].  It's possible, though,
+     * that the handshake has finished and the far end has sent application data but we've not yet set
+     * [dtlsTransport], so we "miss" it.  We don't lose this data, but it will sit inside of [incomingProtocolData]
+     * until the next packet comes through.  This means that we must make a copy of the buffer we receive, as we
+     * won't necessarily be done with it by the time this method completes.
      */
-    override fun send(buf: ByteArray, off: Int, length: Int) {
-        // The buf coming from here will belong to usrsctp, so make sure we copy
-        // into a buffer of our own before forwarding.
-        // NOTE: the reason we have to copy into our own before before forwarding
-        // is that this packet will end up in the same place as a packet that
-        // came from the pool which has to be returned.  Since we can't tell if
-        // a buffer came from the pool or not, we'll need to always return it.
-        val newBuf = BufferPool.getBuffer(length)
-        System.arraycopy(buf, off, newBuf, 0, length)
-        val packet = PacketInfo(UnparsedPacket(newBuf, 0, length))
-        onOutgoingProtocolData(listOf(packet))
+    fun processIncomingProtocolData(data: ByteArray, off: Int, len: Int) {
+        val bufCopy = BufferPool.getBuffer(len).apply {
```

Comment:
The copying in this method looks a bit awkward, but it's effectively the same as what we had before (previously this first copy was done before the stack, but I think it's a bit closer to where we'd like to be to have the stack pull a buffer from the pool itself, that way it doesn't have to assume where the buffer came from)